doc/io_wip.md

```diff
@@ -0,0 +1,206 @@
+buffer.base = Malloc<uint8_t*>(100);
+buffer.len = 100;
+
+io_pull_t pull;
+
+int status = mySource->Pull(&pull, &buffer.len, &buffer, 1, IO_PULL_FLAG_SYNC);
+
+CHECK(status & IO_PULL_STATUS_SYNC);
+if (status & IO_PULL_STATUS_ERROR) {
+  int err = io_get_error(status);
+  // do something with the error
+  return;
+}
+
+// the data should be in buffer
+```
+
+### Example 2: Peek first, then read sync
+
+```c
+Source* mySource = GetSourceSomehow();
+
+size_t length = 100;
+
+io_pull_t pull;
+
+// Peek to get the length
+int status = mySource->Pull(&pull, &len, nullptr, 0,
+                            IO_PULL_FLAG_SYNC |
+                            IO_PULL_FLAG_PEEK);
+
+CHECK(status & IO_PULL_STATUS_SYNC);
+if (status & IO_PULL_STATUS_ERROR) {
+  int err = io_get_error(status);
+  // do something with the error
+  return;
+}
+
+io_buf_t buffer;
+buffer.base = Malloc<uint8_t*>(length);
+buffer.len = length;
+
+status = mySource->Pull(&pull, &length, &buffer, 1, IO_PULL_FLAG_SYNC);
+
+CHECK(status & IO_PULL_STATUS_SYNC);
+if (status & IO_PULL_STATUS_ERROR) {
+  int err = io_get_error(status);
+  // do something with the error
+  return;
+}
+
+// the data should be in buffer
+```
+
+### Example 3: Async pull, data copied into provided buffer
+
+```c
+Source* mySource = GetSourceSomehow();
+
+io_buf_t buffer;
+buffer.base = Malloc<uint8_t*>(100);
+buffer.len = length;
+
+auto AfterPull = [](io_pull_t* handle,
+                    int status,
+                    io_buf_t* bufs,
+                    size_t count,
+                    size_t length,
+                    io_push_t* source) {
+  if (status & IO_PULL_STATUS_ERROR) {
+    int error = io_get_error(status);
+    // handle the error;
+    return;
+  }
+
+  // data should be in bufs
+};
+
+io_pull_t pull;
+io_pull_set_pull_cb(&pull, AfterPull);
+
+int status = mySource->Pull(&pull, &buffer.len, &buffer, 1,
+                            IO_PULL_FLAG_MUST_COPY);
+CHECK(!(status & IO_PULL_STATUS_SYNC));
+```
+
+### Example 4: Async pull, source provided buffer
+
+```c
+Source* mySource = GetSourceSomehow();
+
+io_buf_t buffer;
+buffer.base = Malloc<uint8_t*>(100);
+buffer.len = length;
+
+auto AfterPull = [](io_pull_t* handle,
+                    int status,
+                    io_buf_t* bufs,
+                    size_t count,
+                    size_t length,
+                    io_push_t* source) {
+  if (status & IO_PULL_STATUS_ERROR) {
+    int error = io_get_error(status);
+    // handle the error;
+    return;
+  }
+
+  // data should be in bufs, handle it, then notify source that we're done
+  auto cb = io_push_get_push_cb(source);
+  if (cb != nullptr)
+    cb(source, IO_PUSH_STATUS_OK);
+};
+
+io_pull_t pull;
+io_pull_set_pull_cb(&pull, AfterPull);
+
+int status = mySource->Pull(&pull, &buffer.len, &buffer, 1);
+CHECK(!(status & IO_PULL_STATUS_SYNC));
+```
+
+### Example 5: Binding a Source and Sink
+
+Binding is roughly the equivalent of `pipe()` in Readable, with the notable
```

Comment:
I think I’d prefer to call this ‘piping’ then?

And just to be clear, I understand correctly that only one binding/piping operation is happening for any given source, right?

Reply:
> I think I'd prefer to call this 'piping' then?

I considered that, but given that it is pull based rather than push based, I wanted to avoid accidentally conflating the two.

> ... only one binding/piping operation is happening for any given source, right?

That's the idea but we need to decide that for certain. When we discussed this over dinner in Vancouver, the idea was that data flow would be either One-Source-to-One-Sink, or One-Sink-Many-Sources, in order to keep things as simple as possible.